### PR TITLE
quantum/bitwise.{c,h}: use GCC's __builtin_{clz,popcount}{,l}() when available

### PR DESCRIPTION
**Pull request by liyang**
## Description

Just noticed we're still using our own hand-rolled bit-twiddling code. I didn't see any obvious equivalents for our `bitrev()`.

Should we just remove the `#ifndef`'d-out definitions completely?

```
commit 29e3788044e31b7e6f3e72f973f918296d096d0f (develop/bitwise)
Author: Liyang HU <git@liyang.hu>
Date:   Sun Feb 21 14:50:18 2021 +0000

    quantum/bitwise.{c,h}: use GCC's __builtin_{clz,popcount}{,l}() when available
    
    Available as of GCC 3.4.6, released on 2006-03-06.
    
    https://gcc.gnu.org/onlinedocs/gcc-3.4.6/gcc/Other-Builtins.html

commit 0a52a93b62692e59be650bd6fba1940ad0db4057
Author: Liyang HU <git@liyang.hu>
Date:   Thu Mar 4 02:44:26 2021 +0000

    quantum/bitwise.c: use `Q_NEVER_INLINE`

commit 6a6594bd8752ca765ae8e536f47d52a1d1699fa1
Author: Liyang HU <git@liyang.hu>
Date:   Thu Mar 4 02:44:26 2021 +0000

    quantum/util.h: `#define Q_NEVER_INLINE` & `Q_ALWAYS_INLINE`
    
    ‘Q’ is short for ‘QMK’. Always has been.
```

## Types of Changes

- [x] Core
- [ ] ~~Bugfix~~
- [ ] ~~New feature~~
- [x] Enhancement/optimization
- [ ] ~~Keyboard (addition or update)~~
- [ ] ~~Keymap/layout/userspace (addition or update)~~
- [ ] ~~Documentation~~

## Issues Fixed or Closed by This PR

* ~~n/a~~

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] ~~My change requires a change to the documentation.~~
- [ ] ~~I have updated the documentation accordingly.~~
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] ~~I have added tests to cover my changes.~~
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).
  I've had this change on my own keyboard for the past week.

### CODE REVIEW

**Comment by tzarc** [comment]
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Tested this out and it looks like it is not accurate.   For instance, it breaks my layer code, that uses `biton32` to determine the default layers.  Both the OLED, and the rgb incorrectly identify the layers with this change.

**Comment by liyang** [comment]
> Tested this out and it looks like it is not accurate. For instance, it breaks my layer code, that uses `biton32` to determine the default layers. Both the OLED, and the rgb incorrectly identify the layers with this change.

Oh crap, it's off by one isn't it? Embarrassing. Looks like I forgot about this comment:
```
// NOTE: return 0 when bit0 is on or all bits are off
```
So it should have been `n ? 31 - __builtin_clzl(bits) : 0`, not `32`. My version returned distinct results for `biton(0)` and `biton(1)`.

I was sure _somewhere_ in my firmware one of the `biton*()`s got used, but clearly not! (Or it wasn't used anywhere noticeable.) PR amended. Could you try again?

**Comment by drashna** [comment]
> Oh crap, it's off by one isn't it? Embarrassing. Looks like I forgot about this comment:

No worries. And yup, that was exactly it. 

And yup, that fixes the issue.

**Comment by drashna** [comment]
Correction: Works on AVR.  On ARM, not so much:
```
 |
 | /usr/local/Cellar/arm-gcc-bin@8-2019-q3-update/8-2019-q3-update/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: .build/obj_planck_rev6_drashna/rgb_stuff.o: in function `layer_state_set_rgb_light':
 | /Users/drashna/repos/qmk_firmware/users/drashna/rgb_stuff.c:180: undefined reference to `biton32'
 | /usr/local/Cellar/arm-gcc-bin@8-2019-q3-update/8-2019-q3-update/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: .build/obj_planck_rev6_drashna/encoder_stuff.o: in function `encoder_update_user':
 | /Users/drashna/repos/qmk_firmware/users/drashna/encoder_stuff.c:38: undefined reference to `biton32'
 | collect2: error: ld returned 1 exit status
 |
```

**Comment by liyang** [comment]
Something I don't understand is going on. I did `make planck/rev6:drashna`:
```
QMK Firmware 0.12.8                                                                                     
Making planck/rev6 with keymap drashna                                                                  
                                                                                                        
arm-none-eabi-gcc (15:4.9.3+svn231177-1) 4.9.3 20150529 (prerelease)                                       
… (trimmed) …
Compiling: users/drashna/rgb_stuff.c                                                                [OK]
… (trimmed) …
Linking: .build/planck_rev6_drashna.elf                                                             [OK]
Creating binary load file for flashing: .build/planck_rev6_drashna.bin                              [OK]
Creating load file for flashing: .build/planck_rev6_drashna.hex                                     [OK]

Size after:
   text    data     bss     dec     hex filename
      0   49360       0   49360    c0d0 .build/planck_rev6_drashna.hex

Copying planck_rev6_drashna.bin to qmk_firmware folder                                              [OK]
(Firmware size check does not yet support cortex-m4 microprocessors; skipping.)
```

Looking at my `.build/obj_planck_rev6_drashna/rgb_stuff.lst`, there's no reference to a `biton32` symbol after the ‘bitwise’ patch, but there is before as expected. Your `ld` errors are complaining about a missing `biton32` symbol, which means that they never got inlined into `rgb_stuff.c` (nor `encoder_stuff.c`, which I don't have here).

I think I have misunderstood the meaning of `inline`; I'm a bit rusty with my C. Not sure if I should `__attribute__((always_inline))` or just generate an `extern` symbol for it in `bitwise.c`—guessing either should be fine, but will take a careful read of the [GCC docs](https://gcc.gnu.org/onlinedocs/gcc/Inline.html) (and possibly other references) first.

However I can't replicate the problem in my version of GCC (even with `CFLAGS=-O0`), so… will report back after I try a newer version where I can.

**Comment by drashna** [comment]
Ah, yeah, I'm using 8.3.1 for ARM GCC.  So different version may produce different results...

**Comment by liyang** [comment]
> Ah, yeah, I'm using 8.3.1 for ARM GCC. So different version may produce different results...

I went and installed a copy of ARM GCC, and managed to replicate what you saw.
```
arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 8-2019-q3-update) 8.3.1 20190703 (release) [gcc-8-branch revision 273027]
```

The nub of the problem was indeed that I was using `inline` incorrectly. This [Stack Overflow answer](https://stackoverflow.com/a/6312854/650044) explains it pretty well. Basically by the C99 rules, defining a function `inline` in a header means the compiler is free to inline that definition, but it _can also_ assume the existence of the symbol if it chooses not to inline. (You're supposed to put an `extern inline void foo();` in _some_ `.c` file to generate such a symbol, which I hadn't.) I don't understand why GCC chose _not_ to inline that trivial definition, but—as mentioned—it's free to do so according to the rules.

It just so happens that I've been working on some code that makes use of another project beginning with ‘Q’ recently (available under GPLv3 FWIW), and their approach to forcing inlining is to
```c
#    define Q_NEVER_INLINE __attribute__((noinline))
#    define Q_ALWAYS_INLINE inline __attribute__((always_inline))
```
So marking those in `bitwise.h` with `Q_ALWAYS_INLINE` will ensure things always get inlined by GNU99 rules
https://github.com/qmk/qmk_firmware/blob/b1a8fafa62ec058470715a6930317da09d63181d/tmk_core/rules.mk#L56
and (tested here) resolves the issue you saw.

… except if someone tries to take the address `&biton`. (But why would you do such a thing‽) Then the compiler will emit an external reference to a `biton` symbol anyway. This is taken care of by the `extern Q_ALWAYS_INLINE`s in `bitwise.c` which [causes external symbols to be generated](https://stackoverflow.com/q/216510/650044), and—unless someone does the aforementioned—is dutifully discarded by the linker. From my `.build/planck_rev6_drashna.map`:
```
…
Discarded input sections
…
 .text.bitpop   0x0000000000000000        0xa .build/obj_planck_rev6_drashna/quantum/bitwise.o
 .text.bitpop16 0x0000000000000000        0xa .build/obj_planck_rev6_drashna/quantum/bitwise.o
 .text.bitpop32 0x0000000000000000        0xa .build/obj_planck_rev6_drashna/quantum/bitwise.o
 .text.biton    0x0000000000000000        0xe .build/obj_planck_rev6_drashna/quantum/bitwise.o
 .text.biton16  0x0000000000000000        0xe .build/obj_planck_rev6_drashna/quantum/bitwise.o
 .text.biton32  0x0000000000000000        0xe .build/obj_planck_rev6_drashna/quantum/bitwise.o
…
```

I've tried future-proofing the code a bit with `__has_builtin()` and `__has_attribute()`, although I should note that Clang also (unless told not to) defines `__GNUC__` because that's [a language standard, not a compiler](https://stackoverflow.com/a/55926503/650044). Locally I removed `defined(__GNUC__) ||` from `bitwise.h`:
https://github.com/liyang/qmk_firmware/blob/9b4747f10f16fb9fea4502ab740877b5604ed33b/quantum/bitwise.h#L27
and with GCC 10:
```
arm-none-eabi-gcc (GNU Arm Embedded Toolchain 10-2020-q4-major) 10.2.1 20201103 (release)
```
`biton32()` is still inlined and the `biton32` symbol discarded, according to the resulting `.build/planck_rev6_drashna.map`. In this case with GCC 8, the old hand-rolled `biton32()` gets used, as expected.

I hope that's all the bases covered. 🤞🏾

**Review by drashna — APPROVED** [review]

File: quantum/util.h [hunk 1]
```diff
@@ -16,6 +16,28 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+#ifndef __has_builtin
+#    define __has_builtin(_) 0
+#endif
+
+#ifndef __has_attribute
+#    define __has_attribute(_) 0
+#endif
+
+#if defined(__GNUC__) || __has_attribute(noinline)
+#    define Q_NEVER_INLINE __attribute__((noinline))
```
Comment (fauxpark) [inline quantum/util.h:30]: LUFA already defines these two as `ATTR_NO_INLINE` and `ATTR_ALWAYS_INLINE`:
https://github.com/qmk/lufa/blob/HEAD/LUFA/Common/Attributes.h#L91

We should use those names and fall back to these defines when not compiling a LUFA board.

**Pull request closed without merging**